### PR TITLE
Revise message broker interface

### PR DESCRIPTION
**Pull request by collijk**
Updates the abstract interface for the message broker as it starts to meet the real world in #3969

### CODE REVIEW

File: autogpt/core/command/base.py [hunk 1]
```diff
@@ -1,31 +1,34 @@
 import abc
 import importlib
 import inspect
-from typing import Any, Callable, Optional, Tuple, Dict
```
Comment (collijk) [inline autogpt/core/command/base.py:4]: Everything in this file is from running black.  Sorry!

**Comment by anonhostpi** [comment]
I'm assuming this is related to #3831

File: autogpt/core/messaging/base.py [hunk 2]
```diff
@@ -1,73 +1,193 @@
...
 
 
-class Receiver:
-    id: int  # Who is receiving the message, should use interface receiver
-    role: Role
+@dataclass
+class MessageMetadata:
+    """Struct for metadata about a message."""
+
+    sender: Sender
```
Comment (rihp) [inline autogpt/core/messaging/base.py:31]: i like this approach to cohesively have the sender in the meta data

File: autogpt/core/messaging/base.py [hunk 3]
```diff
@@ -1,73 +1,193 @@
...
+    timestamp: datetime
+    additional_metadata: Dict = None
+    # TODO: what else goes here?
```
Comment (rihp) [inline autogpt/core/messaging/base.py:34]: maybe commands and arguments ?

File: autogpt/core/messaging/base.py [hunk 4]
```diff
@@ -1,73 +1,193 @@
...
+
+
+@dataclass
+class Message:
+    """Struct for a message and its metadata."""
 
+    content: dict  # Some json struct we can define with a strict schema
```
Comment (rihp) [inline autogpt/core/messaging/base.py:41]: Would this be the msg broker contract we have talked about in the past?

Comment (rihp) [reply autogpt/core/messaging/base.py:41]: Where in the code should this structure be defined ?

File: autogpt/core/messaging/base.py [hunk 5]
```diff
@@ -1,73 +1,193 @@
...
-        return []
+        message_channel: "MessageChannel",
+        sender_name: str,
+        sender_role: Role,
+        **additional_metadata,
+    ):
+        ...
+
+    @abc.abstractmethod
+    def send_message(self, content: dict, **extra_metadata) -> None:
```
Comment (rihp) [inline autogpt/core/messaging/base.py]: should there be a return type to confirm that the message was sent? Maybe a boolean?

File: autogpt/core/messaging/base.py [hunk 6]
```diff
@@ -1,73 +1,193 @@
...
+        """Send a message on this channel.
+
+        Args:
+            message: The message to send.
+
+        """
+        ...
+
+    @abc.abstractmethod
+    def __repr__(self) -> str:
```
Comment (rihp) [inline autogpt/core/messaging/base.py:131]: This method could return the metadata from the channel and the most recent message for example

File: autogpt/core/messaging/base.py [hunk 7]
```diff
@@ -1,73 +1,193 @@
...
+                        sent on this channel.
+            message_filter: A function that filters messages before they are sent to
+                              the listener.  If the filter returns False, the message
+                              is not sent to the listener.
+
+        """
+        ...
+
+    @abc.abstractmethod
+    def send_message(self, message: Message) -> None:
```
Comment (rihp) [inline autogpt/core/messaging/base.py]: should this return a boolean?

**Review by rihp — CHANGES_REQUESTED** [review]

File: autogpt/core/messaging/base.py (hunk 3 above)
Comment (collijk) [reply autogpt/core/messaging/base.py:34]: Messages won't always have commands, I think.  The `additional_metadata` dict gives us a lot of flexibility for now until you and @merwanehamadi and others can pull together the data interchange standard

File: autogpt/core/messaging/base.py (hunk 4 above)
Comment (collijk) [reply autogpt/core/messaging/base.py:41]: The content and the metadata. 

They should get define in this subpackage, but I'm not sure what tool you want to use.  There are lots of good ones, though pydantic might be industry standard at this point.  I'm agnostic as long as the schema is enforcable.

File: autogpt/core/messaging/base.py (hunk 5 above)
Comment (collijk) [reply autogpt/core/messaging/base.py]: Good Q.  I'm not sure if the send receipt would be implementation specific.  We can readjust when we do our first real message broker implementation

File: autogpt/core/messaging/base.py (hunk 6 above)
Comment (collijk) [reply autogpt/core/messaging/base.py:131]: `repr`s have to be strings. Concrete classes can determine what that string should be.

File: autogpt/core/messaging/base.py (hunk 5 above)
Comment (collijk) [reply autogpt/core/messaging/base.py]: I'll swap to a bool for now.

**Pull request merged**